feat(css/minifier) Static computation for 'calc' expressions

ghostd opened the pull request:
<!-- Note: CI script will automatically rebase your PR so please do not rebase unless required -->

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**Description:**

This is a first step to fold the "calc" expressions:
* division operator is not (yet) included: i don't know if we want to transform "1/3" into "0,33333"
* other optimizations could follow (operation with infinity, NaN, etc)

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

**BREAKING CHANGE:**

None except the CSS output

<!--
If this PR introduces a breaking change, it must contain a notice for it to be included in the CHANGELOG. Add description or remove entirely if not breaking.

You may need to update `MIGRATION.md` for the breaking changes.
-->

**Related issue (if exists):**
#5860

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/swc-project/swc?pullRequest=5966) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/swc-project/swc?pullRequest=5966) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/swc-project/swc?pullRequest=5966) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/swc-project/swc?pullRequest=5966) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/swc-project/swc?pullRequest=5966) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/swc-project/swc?pullRequest=5966) it.</sub>

ghostd commented:
I needed to add some "copy_" functions because the AST enums does not have the "Copy" trait ; i also tried to add it, but i would need to change JsWord... i guess this limitation will be gone with a fix for #4946

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,846 @@
+    None
+}
+
+fn copy_ident(i: &Ident) -> Ident {
```
You can do `i.clone()` instead

kdy1 replied:
And same for others, too

ghostd replied:
Fixed

kdy1 reviewed (changes requested):
Thanks!

alexander-akait reviewed (changes requested):
Please move **ALL** tests from https://github.com/postcss/postcss-calc/blob/master/test/index.js

I think it can be simplifed, because we have a spec how to calculate it https://www.w3.org/TR/css-values-3/#calc-type-checking

Also current solution only for `calc`, but in spec we have a general solution for all math functions, we should adopt it (AST is almost the same for this functions), but we can do it in another PR

alexander-akait commented:
Anyway let's finish calc firstly (all tests to check everything works fine) and then I will help you to finish other things

ghostd commented:
@alexander-akait i copied most of the tests from postcss-calc except these ones (which depend on an postcss configuration or an exception):

```js
test(
  'should handle precision correctly (3)',
  testValue('calc(5/1000000)', '0.000005', { precision: 6 })
);

test(
  'should preserve the original declaration when `preserve` option is set to true',
  testCss('foo{bar:calc(1rem * 1.5)}', 'foo{bar:1.5rem;bar:calc(1rem * 1.5)}', {
    preserve: true,
  })
);

test(
  'should not yield warnings when nothing is wrong',
  testValue('calc(500px - 0px)', '500px', { warnWhenCannotResolve: true })
);

test(
  'should warn when calc expression cannot be reduced to a single value',
  testValue('calc(100% + 1px)', 'calc(100% + 1px)', {
    warnWhenCannotResolve: true,
  })
);

test(
  'should not parse variables as calc expressions (#35)',
  testCss(
    'foo:nth-child(2n + $var-calc){}',
    'foo:nth-child(2n + $var-calc){}',
    { selectors: true }
  )
);

test(
  'should return the same and not thrown an exception for attribute selectors without a value',
  testCss('button[disabled]{}', 'button[disabled]{}', { selectors: true })
);

test(
  'should reduce calc in media queries when `mediaQueries` option is set to true',
  testCss('@media (min-width:calc(10px+10px)){}', '@media (min-width:20px){}', {
    mediaQueries: true,
  })
);

test(
  'should reduce calc in selectors when `selectors` option is set to true (1)',
  testCss('div[data-size="calc(3*3)"]{}', 'div[data-size="9"]{}', {
    selectors: true,
  })
);

test(
  'should reduce calc in selectors when `selectors` option is set to true (2)',
  testCss('div:nth-child(2n + calc(3*3)){}', 'div:nth-child(2n + 9){}', {
    selectors: true,
  })
);

test(
  'should throw an exception when attempting to divide by zero',
  testThrows('calc(500px/0)', 'calc(500px/0)', 'Cannot divide by zero')
);

test(
  'should throw an exception when attempting to divide by unit (#1)',
  testThrows(
    'calc(500px/2px)',
    'calc(500px/2px)',
    'Cannot divide by "px", number expected'
  )
);

test(
'error with parsing',
  testThrows(
    'calc(10pc + unknown)',
    'calc(10pc + unknown)',
    'Lexical error on line 1: Unrecognized text.\n\n  Erroneous area:\n1: 10pc + unknown\n^.........^'
  )
);

```

Let me know if it correctly understood your purpose.

alexander-akait commented:
This should work (no need option, should work by default):
```
test(
  'should reduce calc in media queries when `mediaQueries` option is set to true',
  testCss('@media (min-width:calc(10px+10px)){}', '@media (min-width:20px){}', {
    mediaQueries: true,
  })
);
```

alexander-akait commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,749 @@
+
+fn fold_calc_sum(calc_sum: &mut CalcSum) {
+    // TODO we should sort the operands by unit to handle this:
+    // "4px + 3% + 5px + 10%" = "9px + 13%"
```
Do we have tests case for this for future improvements?

ghostd replied:
Yes we hate this test: `width:calc((100px - 1em) + (-50px + 1em))`

alexander-akait commented on crates/swc_css_minifier/tests/fixture/compress-calc/ignore-media-query/input.css:
```diff
@@ -0,0 +1,2 @@
+@media (min-width:calc(10px+10px)){
+}
```
To avoid this please add `color: red` for all tests where simple block is empty (maybe we should add option to disable it for better testing, but let's do it late)

ghostd replied:
Fixed

alexander-akait commented on crates/swc_css_minifier/tests/fixture/compress-calc/units/output.min.css:
```diff
@@ -0,0 +1 @@
+.class1{width:calc(1px + 1)}.class2{width:calc(100% - 60px)}.class3{width:calc(100% - -10px)}.class4{width:950ms}.class5{width:calc(99.99%*1/1 - 0rem)}.class6{width:calc(50% + (5em + 5%))}.class7{width:calc((100px - 1em) + (-50px + 1em))}
```
`width:calc(100% - -10px)` -> `width:calc(100% + 10px)`

alexander-akait replied:
`width:calc((100px - 1em) + (-50px + 1em))` -> `width:50px`, but I see `TODO`, so we can postpone it

ghostd replied:
This case is now handled :)

alexander-akait commented on crates/swc_css_minifier/tests/fixture/compress-calc/vendor/output.min.css:
```diff
@@ -0,0 +1 @@
+.class1{width:-webkit-calc(2px)}.class2{width:-WEBKIT-CALC(2px)}.class3{width:-webkit-calc(100% - -webkit-calc(50% + 25px))}.class4{width:-WEBKIT-CALC(1px + -WEBKIT-CALC(2px/2))}.class5{width:-moz-calc(2px)}.class6{width:-webkit-calc(2px)}
```
`width:-webkit-calc(100% - -webkit-calc(50% + 25px))` -> `width:-webkit-calc(50% - 25px)`

ghostd replied:
This case is now handled.

alexander-akait reviewed (commented):
Let's do some improvement (see comments above)

yisibl commented:
```js
test(
  'should throw an exception when attempting to divide by zero',
  testThrows('calc(500px/0)', 'calc(500px/0)', 'Cannot divide by zero')
);
```

I tested that this is valid and equivalent to `calc(infinity * 1px)`, but I didn't find the relevant text for the specification. 
@devSDK, can you help confirm?

DevSDK commented:
Oh sorry, I thought your question is about serialization.

The division of zero is allowed by the css-values-4 specification which is the new version of the spec.

Please check the spec here:

https://www.w3.org/TR/css-values-4/#calc-type-checking

>  Division by zero is possible, which introduces certain complications. [Math functions](https://www.w3.org/TR/css-values-4/#math-function) follow IEEE-754 semantics for these operations:

yisibl commented:
@DevSDK Thank you for your professional answer. My God, allowing division by `0` is so complicated!

ghostd commented:
I implemented the support of calc with the vendor prefix, and added more unit conversions. I think we can be more clever to not lose some precision and factorize some code with the `length.rs` module.

alexander-akait commented on crates/swc_atoms/words.txt:
```diff
@@ -112,8 +114,6 @@ DeviceOrientationEvent
 DocumentFragment
 DocumentType
-dpi
-dppx
```
Let's don't remove them

ghostd replied:
Re-added

alexander-akait commented on crates/swc_css_minifier/tests/fixture/compress-calc/convert-units/output.min.css:
```diff
@@ -1 +1 @@
-.class1{width:calc(1cm + 1px)}.class2{width:calc(1px + 1cm)}.class3{width:20Q}.class4{width:calc(100.9q + 10px)}.class5{width:calc(10px + 100.9q)}.class6{width:calc(10cm + 1px)}.class7{width:calc(1cm + 1px)}.class8{width:calc(10px + 1q)}.class9{width:401q}.class10{width:41q}.class11{width:calc(10in + 1q)}.class12{width:calc(10pt + 1q)}.class13{width:calc(10pc + 1q)}.class14{width:calc(1q + 10px)} ...
+.class1{width:1.0264583333333333cm}.class2{width:38.79527559055118px}.class3{width:20Q}.class4{width:111.48333333333333q}.class5{width:105.33858267716536px}.class6{width:10.026458333333334cm}.class7{width:1.0264583333333333cm}.class8{width:10.94488188976378px}.class9{width:10.025cm}.class10{width:1.025cm}.class11{width:10.009842519685039in}.class12{width:10.708661417322835pt}.class13{width:10.059 ...
```
Sounds like we need option for this, can we implement this?

ghostd replied:
Can you precise what do you have in mind? Something like a "enableMixedUnit" option? "allowSumOfMixedUnit" ?

alexander-akait commented on crates/swc_css_minifier/tests/fixture/compress-calc/media-query/output.min.css:
```diff
@@ -0,0 +1 @@
+@media(min-width:calc(10px + 10px)){.class1{color:red}}
```
Can you investigate why it was not compressed?

ghostd replied:
Here what i figured out: for now, the "calc folding" work with enums like CalcValue::Dimension, CalcValue::Number, and so on. In the case of the "@media" rule, the AST contains other enums ; the Function value is a vector of PreservedToken.

What should we do:
* change the parser to have a Function with a vector of CalcValue enum?
* enhance the calc computation to work with the PreservedToken enum?

yisibl replied:
It will be done here: https://github.com/swc-project/swc/pull/6140

alexander-akait reviewed (commented):
Good job, some improvements and we can move ahead and in future improve it more

kdy1 commented:
Can you rebase?

ghostd commented:
I did a rework to fix some computations and allow more compression.

Do you wish i squash the commits before rebasing ?

jridgewell reviewed (commented):
Careful with accepting this. esbuild had to fix a bug with imprecise calculations because even small fractions can lead to pixel inaccuracy: https://github.com/evanw/esbuild/issues/1821

ghostd commented:
@jridgewell for now i did not implement the division because of this potential loss of precision.

jridgewell commented:
It's not just division: `calc(1cm + 1px)` is computing to `1.0264583333333333cm`, which is imprecise. Matching units could be combined (and even `cm` and `mm` if you convert the output to whole `mm`), but combining others can lead to pixel difference between the browsers interpretation and Rust's.

ghostd commented:
Yes, i think this why @alexander-akait [suggested](https://github.com/swc-project/swc/pull/5966#discussion_r983838520) an option for this kind of transformation.

I'll look into this tomorrow or the next weekend.

ghostd commented:
Actually, the loss precision can happen whatever the operation is done, for example: `calc(100 + (0.5px - 0.4px - 0.1px))` is currently transformed into `calc(100 + -.000000000000000027755575615628914px)` (instead of `100`) because of the floating representation. There is no division involved, only substractions with floating numbers.

What limit should we set:
* an option to disable all arithmetic operations (and the folding would just remove the 0, unnecessary parenthesis, etc) ?
* an option to only allow operations when the number is an integer and the result is also an integer (in the case of floating point, we'd skip the operation) ?
* something smarter?

kdy1 commented:
You don't need to squash :)

ghostd commented:
I added a check after each operation (add, sub, mul) to verify the result accuracy, now `100-75.37`is not reduced anymore to not lose precision.

ghostd commented:
I removed all inaccurate ratios, now we should not have imprecise computations anymore.

There are still a lot of improvement area (remove zeroes, transform `y /x * x` into `y`, handle infinity/nan, reduce some cos/sin/tan expressions, etc)

alexander-akait reviewed (approved):
Okay, let merge, I see these improvements in future:
1. container queries with `calc`
2. `@media`/`@support`, I need investigate why we have list of `PreservedToken` there
3. Some perf improvement, but need look after we fully finished it
4. Better compress in some cases

Anyway I think the current solution is enough to solve popular problems

alexander-akait commented:
@ghostd Also can you solve yarn conflicts?

ghostd commented:
@alexander-akait done :)

kdy1 commented on crates/swc_css_minifier/src/compressor/unit/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    ratio_maps: &'a phf::Map<&'static str, &'static RatioMap>,
+    unit1: &'a str,
+    unit2: &'a str,
+) -> Option<&'a f64> {
```
Let's change the return type to `Option<f64>`

kdy1 commented on crates/swc_css_minifier/src/compressor/unit/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    }
+}
+
+pub fn get_duration_ratio<'a>(unit1: &'a str, unit2: &'a str) -> Option<&'a f64> {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/unit/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    get_ratio(&DURATION_RATIO, unit1, unit2)
+}
+
+pub fn get_frequency_ratio<'a>(unit1: &'a str, unit2: &'a str) -> Option<&'a f64> {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/unit/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    get_ratio(&FREQUENCY_RATIO, unit1, unit2)
+}
+
+pub fn get_absolute_length_ratio<'a>(unit1: &'a str, unit2: &'a str) -> Option<&'a f64> {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/unit/mod.rs:
```diff
@@ -0,0 +1,151 @@
+    get_ratio(&ABSOLUTE_LENGTH_RATIO, unit1, unit2)
+}
+
+pub fn get_resolution_ratio<'a>(unit1: &'a str, unit2: &'a str) -> Option<&'a f64> {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+}
+
+impl CalcSumContext {
+    pub fn new() -> Self {
```
I think we can derive `Default` instead

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        remove_unnecessary_nesting_from_calc_sum(calc_sum);
+    }
+
+    fn nested_fold(&mut self, surrounding_operator: &Option<CalcOperator>, calc_sum: &mut CalcSum) {
```
Please use `Option<&CalcOperator>` instead

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+    }
+
+    fn merge_operators(
+        surrounding_operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn merge_operators(
+        surrounding_operator: &Option<CalcOperator>,
+        nested_operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        }
+    }
+
+    fn reduce(&mut self, operator: &Option<CalcOperator>, operand: &CalcProduct) {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        }
+    }
+
+    fn push(&mut self, operator: &Option<CalcOperator>, operand: &CalcProduct) {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+            .push(CalcProductOrOperator::Product(operand.clone()));
+    }
+
+    fn sum_number(&mut self, operator: &Option<CalcOperator>, operand: &CalcProduct, n: &Number) {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn sum_percentage(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        }
+    }
+
+    fn sum_length(&mut self, operator: &Option<CalcOperator>, operand: &CalcProduct, l: &Length) {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn sum_absolute_length(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn sum_other_length(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        operand: &CalcProduct,
+        l: &Length,
+    ) {
+        let unit = l.unit.value.to_ascii_lowercase().to_string();
```
This is inefficient. Please remove `to_string()`

kdy1 replied:
You can change `other_lengths` to `AHashMap<JsWord, IndexedOperatorAndOperand<Length>>`

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        operand: &CalcProduct,
+        u: &UnknownDimension,
+    ) {
+        let unit = u.unit.value.to_ascii_lowercase().to_string();
```
Same here

kdy1 replied:
Same here, you can change the key type of `unknown_dimension` to `JsWord`

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+    frequency: Option<IndexedOperatorAndOperand<Frequency>>,
+    resolution: Option<IndexedOperatorAndOperand<Resolution>>,
+    flex: Option<IndexedOperatorAndOperand<Flex>>,
+    unknown_dimension: HashMap<String, IndexedOperatorAndOperand<UnknownDimension>>,
```
Please use `JsWord` as the key.

kdy1 replied:
Same here, please use `AHashMap`

kdy1 replied:
Also here

kdy1 replied:
And this

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+    number: Option<IndexedOperatorAndOperand<Number>>,
+    percentage: Option<IndexedOperatorAndOperand<Percentage>>,
+    absolute_length: Option<IndexedOperatorAndOperand<Length>>,
+    other_lengths: HashMap<String, IndexedOperatorAndOperand<Length>>,
```
Please change type to `AHashMap`.
You can import it from `swc_common`

kdy1 replied:
Seems like you missed this

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+fn try_to_extract_into_calc_value(calc_sum: &CalcSum) -> Option<CalcValue> {
+    if calc_sum.expressions.len() == 1 {
+        return match calc_sum.expressions.get(0).unwrap() {
```
`&calc_sum.expressions[0]` seems better than `calc_sum.expressions.get(0).unwrap()`

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+// transform "(simple calc-value)" into "simple calc-value"
+fn remove_unnecessary_nesting_from_calc_sum(calc_sum: &mut CalcSum) {
+    if calc_sum.expressions.len() == 1 {
+        match calc_sum.expressions.get(0).unwrap() {
```
Same here, please use index syntax

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+                if let CalcValueOrOperator::Value(CalcValue::Sum(CalcSum {
+                    expressions: nested_expressions,
+                    span: nested_span,
+                })) = calc_product_expressions.get(0).unwrap()
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+                expressions: calc_product_expressions,
+                ..
+            }) if calc_product_expressions.len() == 1 => {
+                match calc_product_expressions.get(0).unwrap() {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn reduce(&mut self, operator: &Option<CalcOperator>, operand: &CalcProduct) {
+        if operand.expressions.len() == 1 {
+            match operand.expressions.get(0).unwrap() {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+            ComponentValue::Function(Function { name, value, .. })
+                if is_calc_function_name(name) && value.len() == 1 =>
+            {
+                match value.get(0).unwrap() {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+                        expressions: calc_sum_expressions,
+                        ..
+                    }) if calc_sum_expressions.len() == 1 => {
+                        match calc_sum_expressions.get(0).unwrap() {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+                                expressions: calc_product_expressions,
+                                ..
+                            }) if calc_product_expressions.len() == 1 => {
+                                match calc_product_expressions.get(0).unwrap() {
```
Same here

ghostd replied:
My bad, i missed that part. Done now :)

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn sum_frequency(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here, please move `&` into `Option`

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn sum_resolution(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+        }
+    }
+
+    fn sum_flex(&mut self, operator: &Option<CalcOperator>, operand: &CalcProduct, f: &Flex) {
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn sum_unknown_dimension(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn new_indexed_data<T>(
+        &mut self,
+        operator: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+    }
+
+    fn try_to_sum_values(
+        operator1: &Option<IndexedData<CalcOperator>>,
```
Same here

kdy1 replied:
Please change this to `Option<&IndexedData<CalcOperator>>`

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1102 @@
+
+    fn try_to_sum_values(
+        operator1: &Option<IndexedData<CalcOperator>>,
+        operator2: &Option<CalcOperator>,
```
Same here

kdy1 commented on crates/swc_css_minifier/src/compressor/unit/mod.rs:
```diff
@@ -0,0 +1,151 @@
+type RatioMap = phf::Map<&'static str, f64>;
+
+// For length ratio, see https://www.w3.org/TR/css-values-4/#absolute-lengths
+static CM: RatioMap = phf_map! {
```
I expect a match expression of `JsWord` to be faster than `phf`

ghostd commented:
@kdy1 I fixed all your remarks

kdy1 commented on crates/swc_css_minifier/src/compressor/calc_sum.rs:
```diff
@@ -0,0 +1,1083 @@
+        nested_operator: Option<&CalcOperator>,
+    ) -> Option<CalcOperator> {
+        match nested_operator {
+            None => surrounding_operator.map(|so| (*so).clone()),
```
I think `surrounding_operator.cloned()` is better than `surrounding_operator.map(|so| (*so).clone())`

ghostd replied:
You're right, done

kdy1 reviewed (dismissed):
Thank you!

Merging as @alexander-akait  approved, and I verified rust code styles

---

swc-bump:
 - swc_css_minifier

ghostd commented:
It seems i forgot to update some "snapshots" for swc_css_parser... but i don't find an easy way to update the files :-/

kdy1 commented:
`UPDATE=1 cargo test` in the crate directory will do the trick
